clang/test/Frontend/sycl-aspects-usage.cpp

```diff
@@ -0,0 +1,42 @@
+// Test checks Propagate Aspect Usage pass.
+
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm %s -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-WARNINGS
+
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown %s -debug-info-kind=constructor -dwarf-version=5 -emit-llvm -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-WARNINGS-DBG -DPATH=%s
+
+// CHECK-WARNINGS: warning: function 'func1(int)' uses aspect 'fp16' not listed in 'sycl::device_has()'
+// CHECK-WARNINGS-NEXT: note: the actual use is in func3(int, int, int), compile with '-g' to get source location
+// CHECK-WARNINGS-NEXT: note: which is called by func2(int, int), compile with '-g' to get source location
+// CHECK-WARNINGS-NEXT: note: which is called by func1(int), compile with '-g' to get source location
+//
+// CHECK-WARNINGS-DBG: warning: function 'func1(int)' uses aspect 'fp16' not listed in 'sycl::device_has()'
+// CHECK-WARNINGS-DBG-NEXT: note: the actual use is in func3(int, int, int) at [[PATH]]:29:5
+// CHECK-WARNINGS-DBG-NEXT: note: which is called by func2(int, int) at [[PATH]]:33:34
+// CHECK-WARNINGS-DBG-NEXT: note: which is called by func1(int) at [[PATH]]:35:62
+
+#include "Inputs/sycl.hpp"
+
+using namespace cl::sycl;
+
+class KernelName;
+
+struct [[__sycl_detail__::__uses_aspects__(aspect::fp16)]] Struct {
```

Comment:
Added check of empty `device_has` and `double` type. Unfortunately, realized that `__uses_aspects__()` for function's declaration is not implemented yet. I think it would be better to implement it in the following patch.

Reply:
I am sorry. I meant that `__uses_aspects__()` is not handled in the proposed pass at the current moment. On the FE side it is implemented of course.